extensions/kafka/src/main/java/io/deephaven/kafka/publish/PublishToKafka.java

```diff
@@ -89,26 +124,46 @@ public class PublishToKafka<K, V> extends LivenessArtifact {
      * @param valueChunkSerializer Optional {@link KeyOrValueSerializer} to consume table data and produce Kafka record
      *        values in chunk-oriented fashion
      * @param publishInitial If the initial data in {@code table} should be published
+     * @param topicColumn The topic column. When set, uses the the given {@link CharSequence} column from {@code table}
+     *        as the first source for setting the kafka record topic.
+     * @param partitionColumn The partition column. When set, uses the the given {@code int} column from {@code table}
+     *        as the first source for setting the kafka record partition.
+     * @param timestampColumn The timestamp column. When set, uses the the given {@link Instant} column from
+     *        {@code table} as the first source for setting the kafka record timestamp.
      */
     public PublishToKafka(
             final Properties props,
             final Table table,
-            final String topic,
+            final String defaultTopic,
+            final Integer defaultPartition,
             final String[] keyColumns,
             final Serializer<K> kafkaKeySerializer,
             final KeyOrValueSerializer<K> keyChunkSerializer,
             final String[] valueColumns,
             final Serializer<V> kafkaValueSerializer,
             final KeyOrValueSerializer<V> valueChunkSerializer,
+            final ColumnName topicColumn,
+            final ColumnName partitionColumn,
+            final ColumnName timestampColumn,
             final boolean publishInitial) {
         this.table = table;
```

Comment:
Maybe make the RHS of the assignment `.coalesce()`, and then use `this.table` throughout.